src/main/java/com/boot/cart/service/CartService.java

```diff
@@ -1,255 +1,225 @@
+            throw new EntityNotFoundException("Cart not found in the Database!");
         }
 
-        double productTotalRemove = 0;
+        List<CartEntry> cartEntries;
+        Integer initialQuantity = 0;
 
-        List<Product> products = productList.stream().filter(p -> p.getProductName().equals(productName)).collect(Collectors.toList());
-        productList.removeAll(products);
+        cartEntries = cart.getEntries();
+        CartEntry cartEntry = cartEntries.stream().filter(entry -> productName.equals(entry.getProductName())).findFirst().orElse(null);
 
-        cart.setTotal(cart.getTotal() - (products.size() * product.getProductPrice()));
-
-        for (int i = 0; i < quantity; i++) {
-            productList.add(product);
-            productTotalRemove = product.getProductPrice() + productTotalRemove;
+        if (cartEntry == null) {
+            throw new EntityNotFoundException("CartEntry not found in the Database!");
+        } else {
+            initialQuantity = cartEntry.getQuantity();
+            cartEntry.setQuantity(quantity);
+            cartEntry.setCart(cart);
         }
-        cart.setTotal(cart.getTotal() + productTotalRemove);
 
-        cart.setUser(user);
+        cart.setEntries(cartEntries);
 
-        cart.setProductList(productList);
-
-        cartRepository.save(cart);
-
-        return CartMapper.cartEntityToDto(cart);
+        double total = cart.getTotal() - (initialQuantity * productDTO.getPrice()) + (quantity * productDTO.getPrice());
 
+        cart.setTotal(total);
 
+        return cartEntityToDto(cartRepository.save(cart));
     }
 
-    public CartDTO removeProductFromCart(String email, String productName, int quantity)
-            throws InvalidInputDataException, EntityNotFoundException {
+    public CartDTO removeProductFromCart(String email, String productName)
+            throws  EntityNotFoundException {
+        log.info("removeProductFromCart - process started");
 
-        Product product;
+        ProductDTO productDTO;
         try {
-            product = ProductMapper.DtoToProductEntity(productServiceClient.callGetProductByProductName(productName));
+            productDTO = productServiceClient.callGetProductByProductName(productName);
         } catch (HttpClientErrorException.NotFound e) {
             throw new EntityNotFoundException("Product: " + productName + " not found in the Database!");
         }
 
-        User user;
+        UserDTO user;
         try {
-            user = UserMapper.DtoToUserEntity(userServiceClient.callGetUserByEmail(email));
+            user = userServiceClient.callGetUserByEmail(email);
         } catch (HttpClientErrorException.NotFound e) {
             throw new EntityNotFoundException("Email: " + email + " not found in the Database!");
         }
-        Cart cart = cartRepository.findByUser(user);
-        List<Product> productList = cart.getProductList();
+        Cart cart = cartRepository.findByUserId(user.getId());
```

Comment:
you should check also here that the cart is not null

Reply:
Done!